### PR TITLE
fix: replicator Unicode error

### PR DESCRIPTION
**Pull request by Wazzymandias**
## Motivation

- Solana verifications were causing unicode errors in postgres: `unsupported Unicode escape sequence`
- This is because we store certain attributes as JSONB in postgres
- Postgres does not allow null bytes in JSONB
- Solana addresses, claim signatures, and block hashes, due to their cryptographic nature, can be considered arbitrary binary data which can yield byte arrays that have zero values in them
- Ethereum verifications were not getting this error because their values are hex encoded 

## Change Summary

- Convert Solana addresses and block hashes to base58 encoding prior to replicator insertion
- Convert both Ethereum and Solana claim signatures to hex encoding prior to replicator insertion
- Add TODO to create separate replicator process for Solana

## Merge Checklist

_Choose all relevant options below by adding an `x` now or at any time before submitting for review_

- [x] PR title adheres to the [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) standard
- [x] PR has a [changeset](https://github.com/farcasterxyz/hub-monorepo/blob/main/CONTRIBUTING.md#35-adding-changesets)
- [x] PR has been tagged with a change label(s) (i.e. documentation, feature, bugfix, or chore)
- [ ] PR includes [documentation](https://github.com/farcasterxyz/hub-monorepo/blob/main/CONTRIBUTING.md#32-writing-docs) if necessary.
- [x] All [commits have been signed](https://github.com/farcasterxyz/hub-monorepo/blob/main/CONTRIBUTING.md#22-signing-commits)

## Additional Context

If this is a relatively large or complex change, provide more details here that will help reviewers

---

## PR-Codex overview
The focus of this PR is to fix a unicode error in the replicator and enforce hex signatures for verification messages.

### Detailed summary
- Fixed unicode error in replicator
- Added constraint for hex verification message claim signatures
- Updated Dockerfile versions
- Improved error handling in worker.ts
- Added utility functions for hex encoding
- Updated message processing logic for verification messages

> The following files were skipped due to too many changes: `apps/replicator/src/processors/index.ts`

> ✨ Ask PR-Codex anything about this PR by commenting with `/codex {your question}`

### CODE REVIEW

File: apps/replicator/docker-compose.yml [hunk 1]
```diff
@@ -3,6 +3,9 @@ version: '3.9'
 services:
   replicator:
     image: farcasterxyz/replicator:latest
+    build:
+        context: ../../
+        dockerfile: Dockerfile.replicator
```
Comment (sds) [inline apps/replicator/docker-compose.yml:8]: Does `build` conflict with `image`? Or am I misremembering?

File: apps/replicator/src/processors/index.ts [hunk 2]
```diff
@@ -155,13 +155,19 @@ export async function processMessage(
...
+      case MessageType.VERIFICATION_ADD_ETH_ADDRESS: {
         // TODO: add support for multi-protoocl verification
         if (!isVerificationAddAddressMessage(message))
           throw new AssertionError(`Invalid VerificationAddEthAddressMessage: ${message}`);
         log.debug(`Processing VerificationAddEthAddressMessage ${hash} (fid ${fid})`, { fid, hash });
+        const claimSignature = message.data.verificationAddAddressBody.claimSignature;
+        message.data.verificationAddAddressBody.claimSignature =
+          claimSignature.length > 0 && claimSignature[claimSignature.length - 1] === 0
+            ? claimSignature.slice(0, claimSignature.length - 1)
+            : claimSignature;
```
Comment (sds) [inline apps/replicator/src/processors/index.ts:167]: This is an example where Ruby's [`chomp`](https://ruby-doc.org/3.3.0/String.html#method-i-chomp) would be a nice addition to the JS standard library.

File: apps/replicator/src/processors/index.ts [hunk 3]
```diff
@@ -155,13 +155,19 @@ export async function processMessage(
         log.debug(`Processing LinkRemoveMessage ${hash} (fid ${fid})`, { fid, hash });
         await processLinkRemove(message, operation, trx);
         break;
-      case MessageType.VERIFICATION_ADD_ETH_ADDRESS:
+      case MessageType.VERIFICATION_ADD_ETH_ADDRESS: {
```
Comment (sanjayprabhu) [inline apps/replicator/src/processors/index.ts:158]: stray parens?

**Review by sds — APPROVED** [review]
Do we understand why Solana adds a null byte to the signature? Is there a good reason for such behavior?

File: apps/replicator/src/processors/index.ts (hunk 3 above)
Comment (Wazzymandias) [reply apps/replicator/src/processors/index.ts:158]: nah linter complained

I added new variable which it said could leak in the rest of the switch statement so it made me wrap in braces

File: apps/replicator/docker-compose.yml (hunk 1 above)
Comment (Wazzymandias) [reply apps/replicator/docker-compose.yml:8]: AFAICT if I do `docker compose up` it'll use the image but only use build when I do `docker compose up --build`

Comment (sds) [reply apps/replicator/docker-compose.yml:8]: Neat!

**Comment by Wazzymandias** [comment]
> Do we understand why Solana adds a null byte to the signature? Is there a good reason for such behavior?

AFAICT it seems to be standard for some reason for solana signatures. I thought it would be a null terminating byte, but there are null bytes at seemingly random positions of the signature:

```
 index 16 has null byte 64

      at hasNull (index.test.ts:56:17)

  console.log
    verificationAddAddressBody.claimSignature has null byte
 ```
    


    `    index 46 has null byte 63`
    
    
    
    We're inserting the values directly into the database. Neynar isn't seeing this issue because they iterate through the messages and convert the bytes to hex, which resolves this error

File: apps/replicator/src/processors/index.ts [hunk 4]
```diff
@@ -155,13 +159,61 @@ export async function processMessage(
...
+            if (addressBase58.isErr()) {
+              throw new AssertionError(`Invalid address: ${addressBase58.error}`);
+            }
+
+            const addressBytes = base58ToBytes(addressBase58.value);
+            if (addressBytes.isErr()) {
+              throw new AssertionError(`Invalid address: ${addressBytes.error}`);
+            }
+
+            address = addressBytes.value;
```
Comment (sanjayprabhu) [inline apps/replicator/src/processors/index.ts:196]: This gets hex encoded again?

I think it'll be simpler to add a `processVerificationAddSolAddress`

Comment (Wazzymandias) [reply apps/replicator/src/processors/index.ts:196]: > This gets hex encoded again?

No, it's only for solana address and gets base58 encoded

**Review by sanjayprabhu — APPROVED** [review]
Thank you

**Pull request merged**